torch/multiprocessing/spawn.py

```diff
@@ -96,19 +96,28 @@ def __init__(self, processes, error_files):
     def pids(self):
         return [int(process.pid) for process in self.processes]
 
-    def join(self, timeout=None):
+    def _join_procs_with_timeout(self, timeout: float):
+        """Attempt to join all processes with a shared timeout."""
+        end = time.monotonic() + timeout
+        for process in self.processes:
+            time_to_wait = max(0, end - time.monotonic())
+            process.join(time_to_wait)
+
+    def join(self, timeout: float | None = None, grace_period: float | None = None):
         r"""Join one or more processes within spawn context.
 
         Attempt to join one or more processes in this spawn context.
         If one of them exited with a non-zero exit status, this function
-        kills the remaining processes and raises an exception with the cause
-        of the first process exiting.
+        kills the remaining processes (optionally with a grace period)
+        and raises an exception with the cause of the first process exiting.
 
         Returns ``True`` if all processes have been joined successfully,
         ``False`` if there are more processes that need to be joined.
 
         Args:
-            timeout (float): Wait this long before giving up on waiting.
+            timeout: Wait this long before giving up on waiting.
+            grace_period: When some processes fail, wait this long for others
+                to shutdown gracefully before killing them.
```

Comment:
```suggestion
                to shutdown gracefully before terminating them and then killing them after another grace period if they still did not gracefully exit.
```
To be more accurate to the behavior.